# LIBRERIE

In [164]:
import pandas as pd
import os
import FUNCTIONS as fn

#armonizzazione comuni
import re
import unicodedata

In [165]:
# #Per eliminare i limiti degli output
# pd.set_option("display.max_rows", None)
# pd.set_option("display.max_columns", None)

In [166]:
os.getcwd()

'c:\\Users\\eliav\\Desktop\\Road_accidents_analysis\\Notebooks'

# CARICAMENTO DF GREZZI DI ISTAT E SITUAS e SISTEMAZIONE PER SUCCESSIVA COMBINAZIONE DEI DUE 

In [167]:
df_situas_raw = pd.read_csv("../Data/Df_raw/df_situas_raw.csv")
df_situas_raw

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente),TIME_PERIOD,Codice Provincia (Storico)
0,1,1,1,1001,1001,Agliè,NaN,TO,0,0,2574.0,2001,"13,1462",2001,2557,2001,2001,NaN
1,1,1,1,1002,1002,Airasca,NaN,TO,0,0,3554.0,2001,"15,7393",2001,3543,2001,2001,NaN
2,1,1,1,1003,1003,Ala di Stura,NaN,TO,0,0,479.0,2001,"46,3315",2001,480,2001,2001,NaN
3,1,1,1,1004,1004,Albiano d'Ivrea,NaN,TO,0,0,1696.0,2001,"11,7315",2001,1687,2001,2001,NaN
4,1,1,1,1005,1005,Alice Superiore,NaN,TO,0,0,616.0,2001,"7,3796",2001,619,2001,2001,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
192726,5,20,111,111103,111103,Villaputzu,NaN,SU,0,0,4509.0,2021,"181,3947",2024,4368,2024,2024,111.0
192727,5,20,111,111104,111104,Villasalto,NaN,SU,0,0,989.0,2021,"130,3596",2024,903,2024,2024,111.0
192728,5,20,111,111105,111105,Villasimius,NaN,SU,0,0,3705.0,2021,"58,1781",2024,3735,2024,2024,111.0
192729,5,20,111,111106,111106,Villasor,NaN,SU,0,0,6599.0,2021,"86,8137",2024,6534,2024,2024,111.0


In [168]:
df_situas_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 192731 entries, 0 to 192730
Data columns (total 18 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Codice Ripartizione geografica  192731 non-null  int64  
 1   Codice Regione                  192731 non-null  int64  
 2   Codice Provincia/Uts            192731 non-null  int64  
 3   Codice Comune (alfanumerico)    192731 non-null  int64  
 4   Codice Comune (numerico)        192731 non-null  int64  
 5   Comune                          192731 non-null  str    
 6   Comune (dizione straniera)      2911 non-null    str    
 7   Sigla automobilistica           190523 non-null  str    
 8   Capoluogo di Provincia/Uts      192731 non-null  int64  
 9   Capoluogo di Regione            192731 non-null  int64  
 10  Popolazione legale              192722 non-null  float64
 11  Anno Censimento                 192731 non-null  int64  
 12  Superficie (Kmq)           

In [169]:
df_situas_raw["Codice Comune (numerico)"].nunique()

8578

In [170]:
df_situas_raw= df_situas_raw[["Codice Regione", "Codice Provincia/Uts",
                      "Codice Comune (numerico)",
                      "Comune","Capoluogo di Provincia/Uts", "Capoluogo di Regione", "Superficie (Kmq)", 
                      "Popolazione residente", "Anno (Popolazione residente)", "TIME_PERIOD"]]

df_situas_raw

,Codice Regione,Codice Provincia/Uts,Codice Comune (numerico),Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente,Anno (Popolazione residente),TIME_PERIOD
0,1,1,1001,Agliè,0,0,"13,1462",2557,2001,2001
1,1,1,1002,Airasca,0,0,"15,7393",3543,2001,2001
2,1,1,1003,Ala di Stura,0,0,"46,3315",480,2001,2001
3,1,1,1004,Albiano d'Ivrea,0,0,"11,7315",1687,2001,2001
4,1,1,1005,Alice Superiore,0,0,"7,3796",619,2001,2001
...,...,...,...,...,...,...,...,...,...,...
192726,20,111,111103,Villaputzu,0,0,"181,3947",4368,2024,2024
192727,20,111,111104,Villasalto,0,0,"130,3596",903,2024,2024
192728,20,111,111105,Villasimius,0,0,"58,1781",3735,2024,2024
192729,20,111,111106,Villasor,0,0,"86,8137",6534,2024,2024


In [171]:
df_situas_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 192731 entries, 0 to 192730
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype
---  ------                        --------------   -----
 0   Codice Regione                192731 non-null  int64
 1   Codice Provincia/Uts          192731 non-null  int64
 2   Codice Comune (numerico)      192731 non-null  int64
 3   Comune                        192731 non-null  str  
 4   Capoluogo di Provincia/Uts    192731 non-null  int64
 5   Capoluogo di Regione          192731 non-null  int64
 6   Superficie (Kmq)              192731 non-null  str  
 7   Popolazione residente         192731 non-null  int64
 8   Anno (Popolazione residente)  192731 non-null  int64
 9   TIME_PERIOD                   192731 non-null  int64
dtypes: int64(8), str(2)
memory usage: 14.7 MB


In [172]:
df_situas_raw[df_situas_raw["Superficie (Kmq)"].str.contains("\.", regex=True)]["Superficie (Kmq)"]#così mi fa vedere solo la colonna Superficie con i caratteri contenti .
#N.B. ricorda lo \ prima del . altrimenti cercherebbe una concatenazione di parole e non proprio il punto come valore

#regex=False -->Interpreta la stringa come testo letterale, quindi i caratteri speciali sono esattamente quello che sono 
                #a livello letterale. Serve per avere una stringa esatta
#regex=True --> quando vuoi sfruttare i simboli delle espressioni regolari (\s (spazio), \d, ^, $, ., +, *, ecc.) e 
                # non che sia letterale il carattere speciale

5002      1.287,3372
13104     1.287,3371
21204     1.287,3371
29305     1.287,3371
37406     1.287,3371
45507     1.287,3371
53608     1.287,3371
61709     1.287,3371
69707     1.287,3371
77801     1.287,3371
85893     1.287,3895
93985     1.287,3895
102076    1.287,3895
110135    1.287,3895
118183    1.287,3895
126186    1.287,3895
134169    1.287,3895
142128    1.287,3895
150049    1.287,3895
157953    1.287,3716
165854    1.288,1875
173758    1.288,1875
181658    1.288,1875
189554    1.288,1875
Name: Superficie (Kmq), dtype: str

In [173]:
df_situas_raw["Superficie (Kmq)"] = (df_situas_raw["Superficie (Kmq)"].str.replace(".", "", regex=False)
                                        .str.replace(",", ".", regex=False))

#controllo che non ci siano più , e che effettivamente sia stato tutto sostituisto dal.
df_situas_raw[df_situas_raw["Superficie (Kmq)"].str.contains("\,", regex=True)]["Superficie (Kmq)"]

Series([], Name: Superficie (Kmq), dtype: str)

In [174]:
#converto i dtype in float
df_situas_raw["Superficie (Kmq)"] = df_situas_raw["Superficie (Kmq)"].astype(float)

In [175]:
df_situas_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 192731 entries, 0 to 192730
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Codice Regione                192731 non-null  int64  
 1   Codice Provincia/Uts          192731 non-null  int64  
 2   Codice Comune (numerico)      192731 non-null  int64  
 3   Comune                        192731 non-null  str    
 4   Capoluogo di Provincia/Uts    192731 non-null  int64  
 5   Capoluogo di Regione          192731 non-null  int64  
 6   Superficie (Kmq)              192731 non-null  float64
 7   Popolazione residente         192731 non-null  int64  
 8   Anno (Popolazione residente)  192731 non-null  int64  
 9   TIME_PERIOD                   192731 non-null  int64  
dtypes: float64(1), int64(8), str(1)
memory usage: 14.7 MB


In [176]:
#controllo per vedere se per un comune sono associati 24 anni
df_situas_raw[df_situas_raw["Codice Comune (numerico)"] == 1001][["Codice Comune (numerico)", "TIME_PERIOD"]]

,Codice Comune (numerico),TIME_PERIOD
0,1001,2001
8102,1001,2002
16204,1001,2003
24304,1001,2004
32405,1001,2005
40506,1001,2006
48607,1001,2007
56708,1001,2008
64809,1001,2009
72909,1001,2010


In [177]:
#osservo per ogni codice comunale quanti anni differenti vengono registrati
df_situas_raw.groupby("Codice Comune (numerico)")["TIME_PERIOD"].nunique().value_counts()

TIME_PERIOD
24    7488
16     145
9      144
8      126
10     123
15     118
5       80
18      65
14      64
13      57
17      43
6       25
11      25
19      23
7       19
23       8
22       7
12       5
2        4
1        4
4        3
20       2
Name: count, dtype: int64

Qui si vede chiaramente che non per tutti i comuni sono stati registrati i dati per 24 anni, ci sono state variazioni nei codici / comuni

# DF ISTAT

In [178]:
df_istat_raw = pd.read_csv("../Data/Df_raw/df_istat_raw.csv")
df_istat_raw

,DATAFLOW,FREQ,REF_AREA,DATA_TYPE,RESULT,TIME_PERIOD,OBS_VALUE,OBS_STATUS,NOTE_DS,NOTE_REF_AREA,NOTE_DATA_TYPE,NOTE_RESULT,NOTE_TIME_PERIOD,BASE_PER,UNIT_MEAS,UNIT_MULT
0,IT1:41_983(1.0),A,1001,KILLINJ,F,2001,10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,IT1:41_983(1.0),A,1001,KILLINJ,F,2002,10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,IT1:41_983(1.0),A,1001,KILLINJ,F,2003,7,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,IT1:41_983(1.0),A,1001,KILLINJ,F,2004,13,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,IT1:41_983(1.0),A,1001,KILLINJ,F,2005,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
573547,IT1:41_983(1.0),A,111107,ROADACC,9,2020,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
573548,IT1:41_983(1.0),A,111107,ROADACC,9,2021,5,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
573549,IT1:41_983(1.0),A,111107,ROADACC,9,2022,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
573550,IT1:41_983(1.0),A,111107,ROADACC,9,2023,4,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [179]:
df_istat_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 573552 entries, 0 to 573551
Data columns (total 16 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   DATAFLOW          573552 non-null  str    
 1   FREQ              573552 non-null  str    
 2   REF_AREA          573552 non-null  int64  
 3   DATA_TYPE         573552 non-null  str    
 4   RESULT            573552 non-null  str    
 5   TIME_PERIOD       573552 non-null  int64  
 6   OBS_VALUE         573552 non-null  int64  
 7   OBS_STATUS        0 non-null       float64
 8   NOTE_DS           0 non-null       float64
 9   NOTE_REF_AREA     0 non-null       float64
 10  NOTE_DATA_TYPE    0 non-null       float64
 11  NOTE_RESULT       0 non-null       float64
 12  NOTE_TIME_PERIOD  0 non-null       float64
 13  BASE_PER          0 non-null       float64
 14  UNIT_MEAS         0 non-null       float64
 15  UNIT_MULT         0 non-null       float64
dtypes: float64(9), int64(3), str(4)

Unirò i due df (istat e situas) per i codici comunali (ref_area e codice comune numerico) e i rispettivi time_period dopo aver eliminato le colonne che non mi interessano

In [180]:
df_istat_raw= df_istat_raw[["TIME_PERIOD","REF_AREA","DATA_TYPE","RESULT","OBS_VALUE"]]
df_istat_raw

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE
0,2001,1001,KILLINJ,F,10
1,2002,1001,KILLINJ,F,10
2,2003,1001,KILLINJ,F,7
3,2004,1001,KILLINJ,F,13
4,2005,1001,KILLINJ,F,2
...,...,...,...,...,...
573547,2020,111107,ROADACC,9,2
573548,2021,111107,ROADACC,9,5
573549,2022,111107,ROADACC,9,1
573550,2023,111107,ROADACC,9,4


In [181]:
#anni registrati
df_istat_raw["TIME_PERIOD"].unique()

array([2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011,
       2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022,
       2023, 2024])

In [182]:
# quante osservazioni ci sono
df_istat_raw.groupby("DATA_TYPE").size()
#KILLINJ    382368 perchè unendo le osservazioni di morti e feriti fa si che le registrazioni degli incidenti (ROADACC) siano moltiplicate per 2

DATA_TYPE
KILLINJ    382368
ROADACC    191184
dtype: int64

In [183]:
# quanti eventi/persone rappresentano quelle osservazioni
df_istat_raw.groupby("RESULT")["OBS_VALUE"].sum()

print(f"Gli incidenti totali dal 2001 al 2024 sono : {df_istat_raw[df_istat_raw["RESULT"] == "9"]["OBS_VALUE"].sum()}")
print(f"I feriti totali dal 2001 al 2024 sono : {df_istat_raw[df_istat_raw["RESULT"] == "F"]["OBS_VALUE"].sum()}")
print(f"I morti totali dal 2001 al 2024 sono : {df_istat_raw[df_istat_raw["RESULT"] == "M"]["OBS_VALUE"].sum()}")

Gli incidenti totali dal 2001 al 2024 sono : 4779141
I feriti totali dal 2001 al 2024 sono : 6709026
I morti totali dal 2001 al 2024 sono : 101944


In [184]:
df_istat_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 573552 entries, 0 to 573551
Data columns (total 5 columns):
 #   Column       Non-Null Count   Dtype
---  ------       --------------   -----
 0   TIME_PERIOD  573552 non-null  int64
 1   REF_AREA     573552 non-null  int64
 2   DATA_TYPE    573552 non-null  str  
 3   RESULT       573552 non-null  str  
 4   OBS_VALUE    573552 non-null  int64
dtypes: int64(3), str(2)
memory usage: 21.9 MB


In [185]:
print(f"I caratteri duplicati nel dataframe situas grezzo sono : {df_istat_raw.duplicated().sum()}")
print ("="*10)
print(f"I caratteri duplicati nel dataframe istat grezzo sono : {df_istat_raw.duplicated().sum()}")

I caratteri duplicati nel dataframe situas grezzo sono : 0
I caratteri duplicati nel dataframe istat grezzo sono : 0


In [186]:
print(df_istat_raw[["REF_AREA", "TIME_PERIOD"]].dtypes)
print ("="*10)
print(df_situas_raw[["Codice Comune (numerico)" , "TIME_PERIOD"]].dtypes)

REF_AREA       int64
TIME_PERIOD    int64
dtype: object
Codice Comune (numerico)    int64
TIME_PERIOD                 int64
dtype: object


In [187]:
print(f"I codici numerici unici nel df istat sono {df_istat_raw["REF_AREA"].nunique()}")
print ("="*10)
print(f"I codici numerici unici nel df istat sono {df_situas_raw["Codice Comune (numerico)"].nunique()}")

I codici numerici unici nel df istat sono 8578
I codici numerici unici nel df istat sono 8578


In [188]:
#pd.set_option("display.max_rows", None)

prova=df_istat_raw.groupby("RESULT")["TIME_PERIOD"].value_counts()
display(prova.loc["9"]) #ho dovuto mettere loc poichè result è divenuto un indice

TIME_PERIOD
2002    8102
2003    8102
2005    8101
2006    8101
2007    8101
2008    8101
2001    8101
2004    8100
2009    8100
2010    8094
2011    8092
2012    8092
2013    8092
2014    8059
2015    8047
2016    8000
2017    7978
2018    7954
2019    7915
2020    7904
2021    7904
2022    7904
2023    7901
2024    6339
Name: count, dtype: int64

In [189]:
pd.reset_option("display.max_rows")

In [190]:
print(f"I codici numerici unici nel df istat sono {df_istat_raw["TIME_PERIOD"].value_counts()}")
print ("="*10)
print(f"I periodi registrati nel df situas sono {df_situas_raw["TIME_PERIOD"].value_counts()}")

I codici numerici unici nel df istat sono TIME_PERIOD
2002    24306
2003    24306
2001    24303
2005    24303
2006    24303
2007    24303
2008    24303
2004    24300
2009    24300
2010    24282
2011    24276
2012    24276
2013    24276
2014    24177
2015    24141
2016    24000
2017    23934
2018    23862
2019    23745
2020    23712
2021    23712
2022    23712
2023    23703
2024    19017
Name: count, dtype: int64
I periodi registrati nel df situas sono TIME_PERIOD
2001    8102
2002    8102
2004    8101
2005    8101
2006    8101
2007    8101
2008    8101
2003    8100
2009    8100
2010    8094
2011    8092
2012    8092
2013    8090
2014    8057
2015    8046
2016    7998
2017    7978
2018    7954
2019    7914
2021    7904
2022    7904
2020    7903
2023    7900
2024    7896
Name: count, dtype: int64


In [191]:
df_istat_raw.groupby("TIME_PERIOD")["REF_AREA"].nunique()

TIME_PERIOD
2001    8101
2002    8102
2003    8102
2004    8100
2005    8101
2006    8101
2007    8101
2008    8101
2009    8100
2010    8094
2011    8092
2012    8092
2013    8092
2014    8059
2015    8047
2016    8000
2017    7978
2018    7954
2019    7915
2020    7904
2021    7904
2022    7904
2023    7901
2024    6339
Name: REF_AREA, dtype: int64

# MERGE & CLEANING DEI DUE DF GREZZI

In [192]:
df_final_raw= pd.merge(df_istat_raw, df_situas_raw, left_on=["REF_AREA","TIME_PERIOD"], right_on=["Codice Comune (numerico)","TIME_PERIOD"], how="left", indicator=True)

df_final_raw

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Codice Comune (numerico),Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente,Anno (Popolazione residente),_merge
0,2001,1001,KILLINJ,F,10,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2557.0,2001.0,both
1,2002,1001,KILLINJ,F,10,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2538.0,2002.0,both
2,2003,1001,KILLINJ,F,7,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2588.0,2003.0,both
3,2004,1001,KILLINJ,F,13,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2679.0,2004.0,both
4,2005,1001,KILLINJ,F,2,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2674.0,2005.0,both
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,ROADACC,9,2,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1937,2549.0,2020.0,both
573548,2021,111107,ROADACC,9,5,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2536.0,2021.0,both
573549,2022,111107,ROADACC,9,1,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2575.0,2022.0,both
573550,2023,111107,ROADACC,9,4,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2616.0,2023.0,both


In [193]:
#Guardo se qualcosa non è stato matchato (quindi NON BOTH)
df_final_raw["_merge"].value_counts()

_merge
both          572700
left_only        852
right_only         0
Name: count, dtype: int64

In [194]:
left_only=df_final_raw[df_final_raw["_merge"]=="left_only"]

display(left_only)
display(f"Ci sono {left_only.shape[0]} rows che non hanno trovato corrispondenza nel df situas")

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Codice Comune (numerico),Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente,Anno (Popolazione residente),_merge
65103,2019,6064,KILLINJ,F,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
65122,2019,6064,KILLINJ,M,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
65141,2019,6064,ROADACC,9,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
66828,2019,6089,KILLINJ,F,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
66847,2019,6089,KILLINJ,M,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
566295,2016,107022,KILLINJ,M,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
566306,2016,107022,ROADACC,9,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
566317,2016,107023,KILLINJ,F,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only
566328,2016,107023,KILLINJ,M,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,left_only


'Ci sono 852 rows che non hanno trovato corrispondenza nel df situas'

In [195]:
print(f"ID comunali totali di df: {df_final_raw["REF_AREA"].nunique()}")
print("-"*10)
print(f"ID comunali senza match: {left_only["REF_AREA"].nunique()}")

ID comunali totali di df: 8578
----------
ID comunali senza match: 284


Numero di incidenti, feriti e morti a cui corrispondono

In [196]:
left_only.groupby("RESULT")["OBS_VALUE"].sum()

RESULT
9    6187
F    9163
M     130
Name: OBS_VALUE, dtype: int64

**Quanto incidono sul totale di istat, questi dati non matchati?**

In [197]:
tot = df_final_raw.groupby("RESULT")["OBS_VALUE"].sum()
non_match = left_only.groupby("RESULT")["OBS_VALUE"].sum()

risultato = pd.DataFrame({"No match": non_match,
                          "Tot": tot,
                          "% sul totale": (non_match / tot * 100).round(3)})

risultato

,No match,Tot,% sul totale
RESULT,,,
9,6187,4779141,0.129
F,9163,6709026,0.137
M,130,101944,0.128


**Voglio osservare se questi comuni cambiano in qualche modo o sono informazioni mancanti**

In [198]:
#prendo codice di un comune a caso e vedo qual è il primo anno di NON corrispondenza
codice = left_only["REF_AREA"].iloc[0]
anno = left_only["TIME_PERIOD"].iloc[0]

print(codice, anno)

#guardo se esisteva in df situas --->non esisteva
df_situas_raw[(df_situas_raw["Codice Comune (numerico)"] == codice) &
                (df_situas_raw["TIME_PERIOD"] == anno)]

6064 2019


,Codice Regione,Codice Provincia/Uts,Codice Comune (numerico),Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente,Anno (Popolazione residente),TIME_PERIOD


In [199]:
#guardo quel codice preso come esempio fino a che anno è stato registrato
df_situas_raw[df_situas_raw["Codice Comune (numerico)"] == codice][["Codice Comune (numerico)", "Comune", "TIME_PERIOD"]].sort_values("TIME_PERIOD")

,Codice Comune (numerico),Comune,TIME_PERIOD
920,6064,Cuccaro Monferrato,2001
9022,6064,Cuccaro Monferrato,2002
17124,6064,Cuccaro Monferrato,2003
25224,6064,Cuccaro Monferrato,2004
33325,6064,Cuccaro Monferrato,2005
41426,6064,Cuccaro Monferrato,2006
49527,6064,Cuccaro Monferrato,2007
57628,6064,Cuccaro Monferrato,2008
65729,6064,Cuccaro Monferrato,2009
73829,6064,Cuccaro Monferrato,2010


In [200]:
#cerco di capire se Cuccaro Monferrato (preso come esempio) sia stato soppresso oppure unito ad un altro comune
df_situas_raw[df_situas_raw["Comune"].str.contains("CUCCARO MONFERRATO",case=False,na=False)][
    ["Comune","Codice Comune (numerico)","TIME_PERIOD"]].sort_values("TIME_PERIOD")

,Comune,Codice Comune (numerico),TIME_PERIOD
920,Cuccaro Monferrato,6064,2001
9022,Cuccaro Monferrato,6064,2002
17124,Cuccaro Monferrato,6064,2003
25224,Cuccaro Monferrato,6064,2004
33325,Cuccaro Monferrato,6064,2005
41426,Cuccaro Monferrato,6064,2006
49527,Cuccaro Monferrato,6064,2007
57628,Cuccaro Monferrato,6064,2008
65729,Cuccaro Monferrato,6064,2009
73829,Cuccaro Monferrato,6064,2010


La merge è stata effettuata sulla chiave composta (REF_AREA, TIME_PERIOD) e validata tramite validate="many_to_one". Le osservazioni senza corrispondenza (left_only) sono risultate riconducibili quasi esclusivamente a variazioni amministrative dei comuni. 

# CREAZIONE DF FINALE CON SOLO COLONNE DI INTERESSE

In [201]:
df_final_raw.info()

<class 'pandas.DataFrame'>
RangeIndex: 573552 entries, 0 to 573551
Data columns (total 15 columns):
 #   Column                        Non-Null Count   Dtype   
---  ------                        --------------   -----   
 0   TIME_PERIOD                   573552 non-null  int64   
 1   REF_AREA                      573552 non-null  int64   
 2   DATA_TYPE                     573552 non-null  str     
 3   RESULT                        573552 non-null  str     
 4   OBS_VALUE                     573552 non-null  int64   
 5   Codice Regione                572700 non-null  float64 
 6   Codice Provincia/Uts          572700 non-null  float64 
 7   Codice Comune (numerico)      572700 non-null  float64 
 8   Comune                        572700 non-null  str     
 9   Capoluogo di Provincia/Uts    572700 non-null  float64 
 10  Capoluogo di Regione          572700 non-null  float64 
 11  Superficie (Kmq)              572700 non-null  float64 
 12  Popolazione residente         572700 non-

In [202]:
df_final_raw.isna().sum()

TIME_PERIOD                       0
REF_AREA                          0
DATA_TYPE                         0
RESULT                            0
OBS_VALUE                         0
Codice Regione                  852
Codice Provincia/Uts            852
Codice Comune (numerico)        852
Comune                          852
Capoluogo di Provincia/Uts      852
Capoluogo di Regione            852
Superficie (Kmq)                852
Popolazione residente           852
Anno (Popolazione residente)    852
_merge                            0
dtype: int64

In [203]:
df_final_clean= df_final_raw.copy()

df_final_clean

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Codice Comune (numerico),Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente,Anno (Popolazione residente),_merge
0,2001,1001,KILLINJ,F,10,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2557.0,2001.0,both
1,2002,1001,KILLINJ,F,10,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2538.0,2002.0,both
2,2003,1001,KILLINJ,F,7,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2588.0,2003.0,both
3,2004,1001,KILLINJ,F,13,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2679.0,2004.0,both
4,2005,1001,KILLINJ,F,2,1.0,1.0,1001.0,Agliè,0.0,0.0,13.1462,2674.0,2005.0,both
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,ROADACC,9,2,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1937,2549.0,2020.0,both
573548,2021,111107,ROADACC,9,5,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2536.0,2021.0,both
573549,2022,111107,ROADACC,9,1,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2575.0,2022.0,both
573550,2023,111107,ROADACC,9,4,20.0,111.0,111107.0,Villaspeciosa,0.0,0.0,27.1943,2616.0,2023.0,both


In [204]:
# #PRIMA DI ELIMINARE LA COLONNA MERGE MI RICREO IL DF CON SOLO QUELLI CHE SI SONO MATCHATI
# #SI VANNO A PERDERE SOLO 850 RECORD CIRCA SU PIU' DI 500000 

# df_final_clean=df_final_clean[df_final_clean["_merge"] == "both"].copy()

In [205]:
#Elimino merge che non mi serve più, il codice d
df_final_clean=df_final_clean.drop(columns=["_merge","Anno (Popolazione residente)","Codice Comune (numerico)"]) #ho dovuto riscrivere df_final_clean= perchè drop non eliminerebbe 
                                                            #in maniera permanente la colonna, se non volessi salvare il 
                                                            #df come un nuovo df, dovrei mettere drop(columns=["_merge"], inplace=True)

df_final_clean

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente
0,2001,1001,KILLINJ,F,10,1.0,1.0,Agliè,0.0,0.0,13.1462,2557.0
1,2002,1001,KILLINJ,F,10,1.0,1.0,Agliè,0.0,0.0,13.1462,2538.0
2,2003,1001,KILLINJ,F,7,1.0,1.0,Agliè,0.0,0.0,13.1462,2588.0
3,2004,1001,KILLINJ,F,13,1.0,1.0,Agliè,0.0,0.0,13.1462,2679.0
4,2005,1001,KILLINJ,F,2,1.0,1.0,Agliè,0.0,0.0,13.1462,2674.0
...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,ROADACC,9,2,20.0,111.0,Villaspeciosa,0.0,0.0,27.1937,2549.0
573548,2021,111107,ROADACC,9,5,20.0,111.0,Villaspeciosa,0.0,0.0,27.1943,2536.0
573549,2022,111107,ROADACC,9,1,20.0,111.0,Villaspeciosa,0.0,0.0,27.1943,2575.0
573550,2023,111107,ROADACC,9,4,20.0,111.0,Villaspeciosa,0.0,0.0,27.1943,2616.0


In [206]:
#controllo i tipi di dato
df_final_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 573552 entries, 0 to 573551
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   TIME_PERIOD                 573552 non-null  int64  
 1   REF_AREA                    573552 non-null  int64  
 2   DATA_TYPE                   573552 non-null  str    
 3   RESULT                      573552 non-null  str    
 4   OBS_VALUE                   573552 non-null  int64  
 5   Codice Regione              572700 non-null  float64
 6   Codice Provincia/Uts        572700 non-null  float64
 7   Comune                      572700 non-null  str    
 8   Capoluogo di Provincia/Uts  572700 non-null  float64
 9   Capoluogo di Regione        572700 non-null  float64
 10  Superficie (Kmq)            572700 non-null  float64
 11  Popolazione residente       572700 non-null  float64
dtypes: float64(6), int64(3), str(3)
memory usage: 52.5 MB


In [207]:
df_final_clean.isna().sum()

TIME_PERIOD                     0
REF_AREA                        0
DATA_TYPE                       0
RESULT                          0
OBS_VALUE                       0
Codice Regione                852
Codice Provincia/Uts          852
Comune                        852
Capoluogo di Provincia/Uts    852
Capoluogo di Regione          852
Superficie (Kmq)              852
Popolazione residente         852
dtype: int64

In [208]:
#SISTEMO I TIPI DELLE COLONNE

df_final_clean["Codice Regione"] = (df_final_clean["Codice Regione"].astype("Int64"))
# Int64 (con la I maiuscola) permette di avere interi e valori mancanti nella stessa colonna.
df_final_clean["Codice Provincia/Uts"] = (df_final_clean["Codice Provincia/Uts"].astype("Int64"))
df_final_clean["Capoluogo di Provincia/Uts"] = (df_final_clean["Capoluogo di Provincia/Uts"].astype("Int64"))
df_final_clean["Capoluogo di Regione"] = (df_final_clean["Capoluogo di Regione"].astype("Int64"))
df_final_clean["Popolazione residente"] = (df_final_clean["Popolazione residente"].astype("Int64"))

In [209]:
print(df_final_clean.shape) #righe, colonne
print("-"*50)
print(df_final_clean.info())

(573552, 12)
--------------------------------------------------
<class 'pandas.DataFrame'>
RangeIndex: 573552 entries, 0 to 573551
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   TIME_PERIOD                 573552 non-null  int64  
 1   REF_AREA                    573552 non-null  int64  
 2   DATA_TYPE                   573552 non-null  str    
 3   RESULT                      573552 non-null  str    
 4   OBS_VALUE                   573552 non-null  int64  
 5   Codice Regione              572700 non-null  Int64  
 6   Codice Provincia/Uts        572700 non-null  Int64  
 7   Comune                      572700 non-null  str    
 8   Capoluogo di Provincia/Uts  572700 non-null  Int64  
 9   Capoluogo di Regione        572700 non-null  Int64  
 10  Superficie (Kmq)            572700 non-null  float64
 11  Popolazione residente       572700 non-null  Int64  
dtypes: Int64(5), float6

In [210]:
df_final_clean

,TIME_PERIOD,REF_AREA,DATA_TYPE,RESULT,OBS_VALUE,Codice Regione,Codice Provincia/Uts,Comune,Capoluogo di Provincia/Uts,Capoluogo di Regione,Superficie (Kmq),Popolazione residente
0,2001,1001,KILLINJ,F,10,1,1,Agliè,0,0,13.1462,2557
1,2002,1001,KILLINJ,F,10,1,1,Agliè,0,0,13.1462,2538
2,2003,1001,KILLINJ,F,7,1,1,Agliè,0,0,13.1462,2588
3,2004,1001,KILLINJ,F,13,1,1,Agliè,0,0,13.1462,2679
4,2005,1001,KILLINJ,F,2,1,1,Agliè,0,0,13.1462,2674
...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,ROADACC,9,2,20,111,Villaspeciosa,0,0,27.1937,2549
573548,2021,111107,ROADACC,9,5,20,111,Villaspeciosa,0,0,27.1943,2536
573549,2022,111107,ROADACC,9,1,20,111,Villaspeciosa,0,0,27.1943,2575
573550,2023,111107,ROADACC,9,4,20,111,Villaspeciosa,0,0,27.1943,2616


In [211]:
#Rinomino i valori per renderli più chiari
df_final_clean["DATA_TYPE"] = df_final_clean["DATA_TYPE"].replace({"KILLINJ": "morti_e_feriti",
                                                                    "ROADACC": "incidenti_stradali"})

df_final_clean["RESULT"] = df_final_clean["RESULT"].replace({"M": "morti",
                                                            "F": "feriti",
                                                            "9": "incidenti"})

df_final_clean=df_final_clean.rename(columns={"REF_AREA" : "ID_COMUNE", 
                                                              "Codice Provincia/Uts" : "ID_PROVINCIA",
                                                              "Codice Regione" : "ID_REGIONE",
                                                              "Capoluogo di Provincia/Uts": "CAPOLUOGO PROVINCIA",
                                                              "Capoluogo di Regione" : "CAPOLUOGO REGIONE",
                                                              "Comune" : "COMUNE",
                                                              "Popolazione residente" : "RESIDENTI",
                                                              "Superficie (Kmq)" : "SUPERFICIE (KMQ)",})
df_final_clean

,TIME_PERIOD,ID_COMUNE,DATA_TYPE,RESULT,OBS_VALUE,ID_REGIONE,ID_PROVINCIA,COMUNE,CAPOLUOGO PROVINCIA,CAPOLUOGO REGIONE,SUPERFICIE (KMQ),RESIDENTI
0,2001,1001,morti_e_feriti,feriti,10,1,1,Agliè,0,0,13.1462,2557
1,2002,1001,morti_e_feriti,feriti,10,1,1,Agliè,0,0,13.1462,2538
2,2003,1001,morti_e_feriti,feriti,7,1,1,Agliè,0,0,13.1462,2588
3,2004,1001,morti_e_feriti,feriti,13,1,1,Agliè,0,0,13.1462,2679
4,2005,1001,morti_e_feriti,feriti,2,1,1,Agliè,0,0,13.1462,2674
...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,incidenti_stradali,incidenti,2,20,111,Villaspeciosa,0,0,27.1937,2549
573548,2021,111107,incidenti_stradali,incidenti,5,20,111,Villaspeciosa,0,0,27.1943,2536
573549,2022,111107,incidenti_stradali,incidenti,1,20,111,Villaspeciosa,0,0,27.1943,2575
573550,2023,111107,incidenti_stradali,incidenti,4,20,111,Villaspeciosa,0,0,27.1943,2616


In [212]:
#AGGIUNGO IL NOME DELLE REGIONI

regioni = {
    1: "Piemonte",
    2: "Valle d'Aosta",
    3: "Lombardia",
    4: "Trentino-Alto Adige",
    5: "Veneto",
    6: "Friuli-Venezia Giulia",
    7: "Liguria",
    8: "Emilia-Romagna",
    9: "Toscana",
    10: "Umbria",
    11: "Marche",
    12: "Lazio",
    13: "Abruzzo",
    14: "Molise",
    15: "Campania",
    16: "Puglia",
    17: "Basilicata",
    18: "Calabria",
    19: "Sicilia",
    20: "Sardegna"}

df_final_clean["REGIONE"] = df_final_clean["ID_REGIONE"].map(regioni)

# ALTRO CLEANING

In [213]:
df_final_clean

,TIME_PERIOD,ID_COMUNE,DATA_TYPE,RESULT,OBS_VALUE,ID_REGIONE,ID_PROVINCIA,COMUNE,CAPOLUOGO PROVINCIA,CAPOLUOGO REGIONE,SUPERFICIE (KMQ),RESIDENTI,REGIONE
0,2001,1001,morti_e_feriti,feriti,10,1,1,Agliè,0,0,13.1462,2557,Piemonte
1,2002,1001,morti_e_feriti,feriti,10,1,1,Agliè,0,0,13.1462,2538,Piemonte
2,2003,1001,morti_e_feriti,feriti,7,1,1,Agliè,0,0,13.1462,2588,Piemonte
3,2004,1001,morti_e_feriti,feriti,13,1,1,Agliè,0,0,13.1462,2679,Piemonte
4,2005,1001,morti_e_feriti,feriti,2,1,1,Agliè,0,0,13.1462,2674,Piemonte
...,...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,incidenti_stradali,incidenti,2,20,111,Villaspeciosa,0,0,27.1937,2549,Sardegna
573548,2021,111107,incidenti_stradali,incidenti,5,20,111,Villaspeciosa,0,0,27.1943,2536,Sardegna
573549,2022,111107,incidenti_stradali,incidenti,1,20,111,Villaspeciosa,0,0,27.1943,2575,Sardegna
573550,2023,111107,incidenti_stradali,incidenti,4,20,111,Villaspeciosa,0,0,27.1943,2616,Sardegna


**Essendoci stati cambi amministrativi un singolo ID_COMUNE appare solo in una regione?**

In [214]:
check_region = (df_final_clean.groupby("ID_COMUNE").agg({"ID_REGIONE": "nunique",
                                                        "REGIONE": "nunique"}))

check_region[(check_region["ID_REGIONE"] > 1) |(check_region["REGIONE"] > 1)]

,ID_REGIONE,REGIONE
ID_COMUNE,,


Restituendo un Df vuoto, si ha la conferma che non ogni id comune appare in una regione e quindi potrebbe rappresentare una chiave forte su cui costruire i df

**Essendoci stati cambi amministrativi un singolo ID_COMUNE è associato ad un solo nome?**

In [215]:
check_names = (df_final_clean.groupby("ID_COMUNE")["COMUNE"].nunique())
check_names[check_names > 1].sort_values(ascending=False)

ID_COMUNE
4051     2
5014     2
5020     2
5056     2
5077     2
6045     2
6076     2
11020    2
16192    2
17092    2
17189    2
18026    2
18073    2
18122    2
20057    2
20061    2
21052    2
21053    2
21076    2
21102    2
22029    2
22156    2
22162    2
22176    2
23014    2
23030    2
23052    2
31003    2
31019    2
31022    2
32001    2
32002    2
32005    2
41035    2
65025    2
68033    2
Name: COMUNE, dtype: int64

In [216]:
print("ID con più di un nome:", (check_names > 1).sum())
print("Cambi di nome:", (check_names[check_names > 1] - 1).sum())

ID con più di un nome: 36
Cambi di nome: 36


**Questo mi mostra che per ID_COMUNE a volte ci sono due nomi, ora andrò a conntrollare quali sono**

In [217]:
df_final_clean[df_final_clean["ID_COMUNE"].isin(check_names[check_names > 1].index)& (df_final_clean["RESULT"] == "incidenti")][
    ["ID_COMUNE", "COMUNE", "ID_REGIONE", "REGIONE", "TIME_PERIOD","RESULT"]].sort_values(["ID_COMUNE", "TIME_PERIOD"]).head(25)

,ID_COMUNE,COMUNE,ID_REGIONE,REGIONE,TIME_PERIOD,RESULT
38251,4051,Castellinaldo,1,Piemonte,2001,incidenti
38252,4051,Castellinaldo,1,Piemonte,2002,incidenti
38253,4051,Castellinaldo,1,Piemonte,2003,incidenti
38254,4051,Castellinaldo,1,Piemonte,2004,incidenti
38255,4051,Castellinaldo,1,Piemonte,2005,incidenti
38256,4051,Castellinaldo,1,Piemonte,2006,incidenti
38257,4051,Castellinaldo,1,Piemonte,2007,incidenti
38258,4051,Castellinaldo,1,Piemonte,2008,incidenti
38259,4051,Castellinaldo,1,Piemonte,2009,incidenti
38260,4051,Castellinaldo,1,Piemonte,2010,incidenti


Il caso di ID 4051 è differente da quello di 6064, infatti nonostante il nome cambiato, l' ID rimane invariato, mentre 6064 cambia sia il nome che ID.

Qui è evidente come i comuni cambino i nomi negli anni

In [218]:
#cambio di nome ma stesso ID
cambi_nome = (
    df_final_clean
    .groupby("ID_COMUNE")["COMUNE"]
    .unique()
)

cambi_nome = cambi_nome[cambi_nome.apply(len) > 1]

cambi_nome

ID_COMUNE
4051      [Castellinaldo, Castellinaldo d'Alba]
5014            [Calliano, Calliano Monferrato]
5020              [Casorzo, Casorzo Monferrato]
5056                  [Grana, Grana Monferrato]
5077        [Montemagno, Montemagno Monferrato]
                          ...                  
107019                  [Sant'Anna Arresi, nan]
107020                      [Sant'Antioco, nan]
107021                         [Tratalias, nan]
107022                    [Villamassargia, nan]
107023                     [Villaperuccio, nan]
Name: COMUNE, Length: 320, dtype: object

## Essendoci stati cambi amministrativi però voglio capire se ID_COMUNE non cambi più volte per un comune

In [219]:
df_final_clean.groupby(
    ["REGIONE", "COMUNE"]
)["ID_COMUNE"].nunique().sort_values(ascending=False).head(20)

REGIONE   COMUNE         
Sardegna  Buggerru           3
          Collinas           3
          Carloforte         3
          Calasetta          3
          Carbonia           3
          Las Plassas        3
          Budoni             3
          Arbus              3
          Barumini           3
          Domusnovas         3
          Escalaplano        3
          Escolca            3
          Esterzili          3
          Fluminimaggiore    3
          Furtei             3
          Genoni             3
          Genuri             3
          Gergei             3
          Gesturi            3
          Giba               3
Name: ID_COMUNE, dtype: int64

Questo mostra come per alcuni comuni l' ID cambia diverse volte, nel df che andrò a creare nei prossimi notebook, utilizzerò l' ID più recente

Ora voglio controllare come sono cambiati gli ID, prendendo come esempio Esterzilli

In [220]:
df_final_clean[
    df_final_clean["COMUNE"].isin(["Esterzili"])
][
    ["REGIONE", "COMUNE", "ID_COMUNE", "TIME_PERIOD"]
].drop_duplicates().sort_values(
    ["REGIONE", "COMUNE", "TIME_PERIOD"]
).head(20)

,REGIONE,COMUNE,ID_COMUNE,TIME_PERIOD
512049,Sardegna,Esterzili,91022,2001
512050,Sardegna,Esterzili,91022,2002
512051,Sardegna,Esterzili,91022,2003
512052,Sardegna,Esterzili,91022,2004
512053,Sardegna,Esterzili,91022,2005
519954,Sardegna,Esterzili,92112,2006
519955,Sardegna,Esterzili,92112,2007
519956,Sardegna,Esterzili,92112,2008
519957,Sardegna,Esterzili,92112,2009
519958,Sardegna,Esterzili,92112,2010


Dato che teoricamente il nome di un comune dovrebbe essere associato ad una sola regione, guardo se l' unione di queste due chiavi possa essere più pertinente rispetto a utilizzare l' ID, che come ho osservato non è unicovo per comune

In [221]:
check = (
    df_final_clean
    .groupby(["REGIONE", "COMUNE", "TIME_PERIOD"])["ID_COMUNE"]
    .nunique()
)

check[check > 1]

Series([], Name: ID_COMUNE, dtype: int64)

Non essendoci risultati si può pensare di costruire un df aggregato basandosi sul nome del comune associato alla regione

In [222]:
df_final_clean.groupby(["REGIONE", "COMUNE"])["ID_COMUNE"].nunique().value_counts()

ID_COMUNE
1    7990
2     210
3      68
Name: count, dtype: int64

Ci sono 278 comuni che hanno subito cambi di ID

**Creo un ulteriore chiave univoca, avendo visto che cambiano gli ID_COMUNE negli anni, terrò quello del 2024 e lo propagherò per tutti gli anni precedenti, così da avere un solo ID per tutti gli anni e per quel comune**

In [223]:
df_final_clean[
    df_final_clean["COMUNE"].str.contains(
        "Esterzili", case=False, na=False
    )
][
    ["TIME_PERIOD", "ID_COMUNE", "COMUNE"]
].drop_duplicates().sort_values("TIME_PERIOD")

,TIME_PERIOD,ID_COMUNE,COMUNE
512049,2001,91022,Esterzili
512050,2002,91022,Esterzili
512051,2003,91022,Esterzili
512052,2004,91022,Esterzili
512053,2005,91022,Esterzili
519954,2006,92112,Esterzili
519955,2007,92112,Esterzili
519956,2008,92112,Esterzili
519957,2009,92112,Esterzili
519958,2010,92112,Esterzili


In [224]:
#mantengo comunque ID originale
df_final_clean["ID_COMUNE_OG"] = df_final_clean["ID_COMUNE"]

In [225]:
#creo la corrispondenza comune + regione = id più recente
#prendo i comuni e le regioni non nulli
#ordino in base agli anni, elimino i duplicati comune+regione e mantengo l'ultimo (così da avere id aggiornato)

map_id = (
    df_final_clean[
        df_final_clean["COMUNE"].notna() &
        df_final_clean["REGIONE"].notna()
    ]
    .sort_values("TIME_PERIOD")
    .drop_duplicates(
        subset=["COMUNE", "REGIONE"],
        keep="last"
    )
    [["COMUNE", "REGIONE", "ID_COMUNE"]]
    .rename(columns={"ID_COMUNE": "ID_COMUNE_NUOVO"})
)

In [226]:
df_final_clean = df_final_clean.drop(columns="ID_COMUNE")

df_final_clean = df_final_clean.merge(map_id,on=["COMUNE", "REGIONE"],how="left",validate="many_to_one")

df_final_clean.rename(columns={"ID_COMUNE_NUOVO": "ID_COMUNE"},inplace=True)

In [227]:
df_final_clean[
    df_final_clean["COMUNE"].str.contains(
        "Esterzili", case=False, na=False
    )
]

,TIME_PERIOD,DATA_TYPE,RESULT,OBS_VALUE,ID_REGIONE,ID_PROVINCIA,COMUNE,CAPOLUOGO PROVINCIA,CAPOLUOGO REGIONE,SUPERFICIE (KMQ),RESIDENTI,REGIONE,ID_COMUNE_OG,ID_COMUNE
512049,2001,morti_e_feriti,feriti,0,20,91,Esterzili,0,0,100.7428,836,Sardegna,91022,111020.0
512050,2002,morti_e_feriti,feriti,0,20,91,Esterzili,0,0,100.7428,832,Sardegna,91022,111020.0
512051,2003,morti_e_feriti,feriti,0,20,91,Esterzili,0,0,100.7428,822,Sardegna,91022,111020.0
512052,2004,morti_e_feriti,feriti,3,20,91,Esterzili,0,0,100.7428,820,Sardegna,91022,111020.0
512053,2005,morti_e_feriti,feriti,1,20,91,Esterzili,0,0,100.7428,821,Sardegna,91022,111020.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
571513,2020,incidenti_stradali,incidenti,0,20,111,Esterzili,0,0,100.7428,575,Sardegna,111020,111020.0
571514,2021,incidenti_stradali,incidenti,1,20,111,Esterzili,0,0,100.7430,565,Sardegna,111020,111020.0
571515,2022,incidenti_stradali,incidenti,2,20,111,Esterzili,0,0,100.7430,563,Sardegna,111020,111020.0
571516,2023,incidenti_stradali,incidenti,0,20,111,Esterzili,0,0,100.7430,560,Sardegna,111020,111020.0


In [228]:
df_final_clean.columns.tolist()

['TIME_PERIOD',
 'DATA_TYPE',
 'RESULT',
 'OBS_VALUE',
 'ID_REGIONE',
 'ID_PROVINCIA',
 'COMUNE',
 'CAPOLUOGO PROVINCIA',
 'CAPOLUOGO REGIONE',
 'SUPERFICIE (KMQ)',
 'RESIDENTI',
 'REGIONE',
 'ID_COMUNE_OG',
 'ID_COMUNE']

In [229]:
#Sistemo ordine delle colonne
df_final_clean = df_final_clean[["TIME_PERIOD",
                                 "ID_COMUNE_OG",
                                "ID_COMUNE",
                                "COMUNE",
                                "ID_REGIONE",
                                "REGIONE",
                                "CAPOLUOGO REGIONE",
                                "ID_PROVINCIA",
                                "CAPOLUOGO PROVINCIA",
                                "DATA_TYPE",
                                "RESULT",
                                "OBS_VALUE",
                                "RESIDENTI",
                                "SUPERFICIE (KMQ)"]]

df_final_clean

,TIME_PERIOD,ID_COMUNE_OG,ID_COMUNE,COMUNE,ID_REGIONE,REGIONE,CAPOLUOGO REGIONE,ID_PROVINCIA,CAPOLUOGO PROVINCIA,DATA_TYPE,RESULT,OBS_VALUE,RESIDENTI,SUPERFICIE (KMQ)
0,2001,1001,1001.0,Agliè,1,Piemonte,0,1,0,morti_e_feriti,feriti,10,2557,13.1462
1,2002,1001,1001.0,Agliè,1,Piemonte,0,1,0,morti_e_feriti,feriti,10,2538,13.1462
2,2003,1001,1001.0,Agliè,1,Piemonte,0,1,0,morti_e_feriti,feriti,7,2588,13.1462
3,2004,1001,1001.0,Agliè,1,Piemonte,0,1,0,morti_e_feriti,feriti,13,2679,13.1462
4,2005,1001,1001.0,Agliè,1,Piemonte,0,1,0,morti_e_feriti,feriti,2,2674,13.1462
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
573547,2020,111107,111107.0,Villaspeciosa,20,Sardegna,0,111,0,incidenti_stradali,incidenti,2,2549,27.1937
573548,2021,111107,111107.0,Villaspeciosa,20,Sardegna,0,111,0,incidenti_stradali,incidenti,5,2536,27.1943
573549,2022,111107,111107.0,Villaspeciosa,20,Sardegna,0,111,0,incidenti_stradali,incidenti,1,2575,27.1943
573550,2023,111107,111107.0,Villaspeciosa,20,Sardegna,0,111,0,incidenti_stradali,incidenti,4,2616,27.1943


In [230]:
#per ogni anno quanti ID comunali originali sono presenti?

df_final_clean.groupby("TIME_PERIOD")["ID_COMUNE_OG"].nunique()

TIME_PERIOD
2001    8101
2002    8102
2003    8102
2004    8100
2005    8101
2006    8101
2007    8101
2008    8101
2009    8100
2010    8094
2011    8092
2012    8092
2013    8092
2014    8059
2015    8047
2016    8000
2017    7978
2018    7954
2019    7915
2020    7904
2021    7904
2022    7904
2023    7901
2024    6339
Name: ID_COMUNE_OG, dtype: int64

In [231]:
#per ogni anno quanti ID comunali (propagati) sono presenti?

df_final_clean.groupby("TIME_PERIOD")["ID_COMUNE"].nunique()

TIME_PERIOD
2001    8101
2002    8102
2003    8099
2004    8100
2005    8101
2006    8101
2007    8101
2008    8101
2009    7988
2010    8094
2011    8092
2012    8092
2013    8088
2014    8056
2015    8046
2016    7843
2017    7978
2018    7954
2019    7913
2020    7903
2021    7904
2022    7904
2023    7900
2024    6339
Name: ID_COMUNE, dtype: int64

La colonna ID_COMUNI farà sì che un comune il cui ID sia cambiato diverse volte negli anni, sia stato aggiornato all' ultimo ID assegnato.

Verrà mantenuta comunque la colonna con gli ID originali (*OG*) in maniera tale da poter vedere la mutazione temporale avvenuta per questi ultimi.

In [232]:
check_id = (
    df_final_clean
    .groupby("ID_COMUNE")["COMUNE"]
    .nunique()
)

check_id[check_id > 1]

ID_COMUNE
4051.0     2
5014.0     2
5020.0     2
5056.0     2
5077.0     2
6045.0     2
6076.0     2
11020.0    2
17092.0    2
17189.0    2
18026.0    2
18073.0    2
18122.0    2
20057.0    2
20061.0    2
21052.0    2
21053.0    2
21076.0    2
21102.0    2
22029.0    2
22156.0    2
22162.0    2
22176.0    2
23014.0    2
23030.0    2
23052.0    2
31003.0    2
31019.0    2
31022.0    2
32001.0    2
32002.0    2
32005.0    2
41035.0    2
65025.0    2
68033.0    2
Name: COMUNE, dtype: int64

In [233]:
check_fusioni = (
    df_final_clean
    .groupby("ID_COMUNE")["ID_COMUNE_OG"]
    .nunique()
)

check_fusioni[check_fusioni > 1].sort_values(ascending=False)

ID_COMUNE
111101.0    3
111100.0    3
111099.0    3
111098.0    3
111097.0    3
           ..
90041.0     2
90047.0     2
90049.0     2
90054.0     2
90062.0     2
Name: ID_COMUNE_OG, Length: 278, dtype: int64

In [234]:
df_final_clean[
    df_final_clean["ID_COMUNE"] == 18002][[
        "TIME_PERIOD", "ID_COMUNE_OG", "ID_COMUNE", "COMUNE", "REGIONE"]
        ].drop_duplicates().sort_values("TIME_PERIOD")

,TIME_PERIOD,ID_COMUNE_OG,ID_COMUNE,COMUNE,REGIONE
166092,2001,18002,18002.0,Albaredo Arnaboldi,Lombardia
166093,2002,18002,18002.0,Albaredo Arnaboldi,Lombardia
166094,2003,18002,18002.0,Albaredo Arnaboldi,Lombardia
166095,2004,18002,18002.0,Albaredo Arnaboldi,Lombardia
166096,2005,18002,18002.0,Albaredo Arnaboldi,Lombardia
166097,2006,18002,18002.0,Albaredo Arnaboldi,Lombardia
166098,2007,18002,18002.0,Albaredo Arnaboldi,Lombardia
166099,2008,18002,18002.0,Albaredo Arnaboldi,Lombardia
166100,2009,18002,18002.0,Albaredo Arnaboldi,Lombardia
166101,2010,18002,18002.0,Albaredo Arnaboldi,Lombardia


In [235]:
df_final_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 573552 entries, 0 to 573551
Data columns (total 14 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   TIME_PERIOD          573552 non-null  int64  
 1   ID_COMUNE_OG         573552 non-null  int64  
 2   ID_COMUNE            572700 non-null  float64
 3   COMUNE               572700 non-null  str    
 4   ID_REGIONE           572700 non-null  Int64  
 5   REGIONE              572700 non-null  str    
 6   CAPOLUOGO REGIONE    572700 non-null  Int64  
 7   ID_PROVINCIA         572700 non-null  Int64  
 8   CAPOLUOGO PROVINCIA  572700 non-null  Int64  
 9   DATA_TYPE            573552 non-null  str    
 10  RESULT               573552 non-null  str    
 11  OBS_VALUE            573552 non-null  int64  
 12  RESIDENTI            572700 non-null  Int64  
 13  SUPERFICIE (KMQ)     572700 non-null  float64
dtypes: Int64(5), float64(2), int64(3), str(4)
memory usage: 64.0 MB


In [236]:
df_final_clean["ID_COMUNE"]=df_final_clean["ID_COMUNE"].astype("Int64")

## Ho visto che ci sono state delle fusioni tra comuni, questo ne influenza la superficie? In che modo?

In [237]:
# Controllo della variazione della superficie nel tempo
municipalities_changed_name = check_names[check_names > 1].index

surface_changes = (
    df_final_clean[
        df_final_clean["ID_COMUNE"].isin(municipalities_changed_name)].groupby("ID_COMUNE")["SUPERFICIE (KMQ)"]
    .nunique())

# Visualizza solo i comuni per i quali la superficie è cambiata
surface_changes[surface_changes > 1].sort_values(ascending=False).value_counts()

SUPERFICIE (KMQ)
3    14
2    14
4     4
Name: count, dtype: int64

La superficie è cambiata per alcuni ID comunali 4, 3 oppure 2 volte

**Quali sono questi comuni**

In [238]:
# ID dei comuni per i quali la superficie cambia nel tempo
mun_changed_sup = surface_changes[surface_changes > 1].index

df_final_clean[
    df_final_clean["ID_COMUNE"].isin(mun_changed_sup)][["ID_COMUNE",
                                                        "COMUNE",
                                                        "TIME_PERIOD",
                                                        "SUPERFICIE (KMQ)",
                                                        "RESIDENTI"]].sort_values(
                                                        ["ID_COMUNE", "TIME_PERIOD"])

,ID_COMUNE,COMUNE,TIME_PERIOD,SUPERFICIE (KMQ),RESIDENTI
38205,4051,Castellinaldo,2001,7.8029,861
38228,4051,Castellinaldo,2001,7.8029,861
38251,4051,Castellinaldo,2001,7.8029,861
38206,4051,Castellinaldo,2002,7.8028,871
38229,4051,Castellinaldo,2002,7.8028,871
...,...,...,...,...,...
412357,68033,Popoli Terme,2023,35.0394,4674
412381,68033,Popoli Terme,2023,35.0394,4674
412334,68033,Popoli Terme,2024,35.0394,4620
412358,68033,Popoli Terme,2024,35.0394,4620


In alcuni comuni si può notare come la superficie sia cambiata per un piccolo valore

**Di quanto cambia la superficie?**

In [239]:
df_final_clean[df_final_clean["ID_COMUNE"].isin(mun_changed_sup)][["ID_COMUNE","COMUNE",
                                                                "TIME_PERIOD",
                                                                "SUPERFICIE (KMQ)",
                                                                "RESIDENTI"]].sort_values(["ID_COMUNE", "TIME_PERIOD"])


,ID_COMUNE,COMUNE,TIME_PERIOD,SUPERFICIE (KMQ),RESIDENTI
38205,4051,Castellinaldo,2001,7.8029,861
38228,4051,Castellinaldo,2001,7.8029,861
38251,4051,Castellinaldo,2001,7.8029,861
38206,4051,Castellinaldo,2002,7.8028,871
38229,4051,Castellinaldo,2002,7.8028,871
...,...,...,...,...,...
412357,68033,Popoli Terme,2023,35.0394,4674
412381,68033,Popoli Terme,2023,35.0394,4674
412334,68033,Popoli Terme,2024,35.0394,4620
412358,68033,Popoli Terme,2024,35.0394,4620


In [240]:
surface_variation = (df_final_clean[df_final_clean["ID_COMUNE"].isin(municipalities_changed_name)][
                                            ["ID_COMUNE", "COMUNE", "TIME_PERIOD", "SUPERFICIE (KMQ)"]]
                                                .sort_values(["ID_COMUNE", "TIME_PERIOD"]))

# Superficie minima e massima per ogni comune
surface_range = (surface_variation.groupby("ID_COMUNE").agg(SUPERFICIE_MIN=("SUPERFICIE (KMQ)", "min")
                                                            ,SUPERFICIE_MAX=("SUPERFICIE (KMQ)", "max")))

# Differenza assoluta
surface_range["DIFFERENZA_KMQ"] = (surface_range["SUPERFICIE_MAX"]- surface_range["SUPERFICIE_MIN"])

surface_range.sort_values("DIFFERENZA_KMQ",ascending=False).head(10)

,SUPERFICIE_MIN,SUPERFICIE_MAX,DIFFERENZA_KMQ
ID_COMUNE,,,
20057,24.5304,51.5351,27.0047
20061,57.0440,79.8282,22.7842
18026,3.4926,12.4432,8.9506
65025,113.0263,113.5471,0.5208
21102,20.5608,20.6864,0.1256
21052,46.4363,46.5542,0.1179
21076,33.1250,33.2044,0.0794
21053,19.5102,19.5522,0.0420
23052,40.4219,40.4501,0.0282


Nella maggior parte dei cambi di denominazione la superficie rimane sostanzialmente stabile; solo 2 comuni cambiano classe dimensionale sulla base della superficie minima e massima osservate.

-Classificazione Comuni per superficie
https://www.istat.it/it/files/2015/04/Superfici-delle-unit%C3%A0-amministrative-Testo-integrale.pdf


In [241]:
df_final_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 573552 entries, 0 to 573551
Data columns (total 14 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   TIME_PERIOD          573552 non-null  int64  
 1   ID_COMUNE_OG         573552 non-null  int64  
 2   ID_COMUNE            572700 non-null  Int64  
 3   COMUNE               572700 non-null  str    
 4   ID_REGIONE           572700 non-null  Int64  
 5   REGIONE              572700 non-null  str    
 6   CAPOLUOGO REGIONE    572700 non-null  Int64  
 7   ID_PROVINCIA         572700 non-null  Int64  
 8   CAPOLUOGO PROVINCIA  572700 non-null  Int64  
 9   DATA_TYPE            573552 non-null  str    
 10  RESULT               573552 non-null  str    
 11  OBS_VALUE            573552 non-null  int64  
 12  RESIDENTI            572700 non-null  Int64  
 13  SUPERFICIE (KMQ)     572700 non-null  float64
dtypes: Int64(6), float64(1), int64(3), str(4)
memory usage: 64.5 MB


# SALVATAGGIO DEI DF

In [242]:
os.getcwd()
os.listdir()

['00_fetching.ipynb',
 '01cleaning_overview.ipynb',
 '02a.EDA_&_Temporal_analysis.ipynb',
 '02b_EDA_Municipality_analysis_aggregate_01_24.ipynb',
 '03b_Test.ipynb',
 'FUNCTIONS.py',
 '__pycache__']

In [243]:
if not os.path.exists("../Data/Df_clean"):
    os.mkdir("../Data/Df_clean")
df_final_clean.to_csv("../Data/Df_clean/df_final_from_01_to_24_clean.csv",index=False)

In [244]:
df_incidenti= df_final_clean[df_final_clean["RESULT"]=="incidenti"]
df_incidenti.to_csv("../Data/Df_clean/df_incidenti_from_01_to_24.csv",index=False)

df_morti= df_final_clean[df_final_clean["RESULT"]=="morti"]
df_morti.to_csv("../Data/Df_clean/df_morti_from_01_to_24.csv",index=False)

df_feriti= df_final_clean[df_final_clean["RESULT"]=="feriti"]
df_feriti.to_csv("../Data/Df_clean/df_feriti_from_01_to_24.csv",index=False)